### ISSUE TITLE
commitlog hard limit (almost?) reached with tablets

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylladb#17086]
![image](https://github.com/scylladb/scylladb/assets/1017210/94f09758-4434-4c73-ab45-5e1128330980)

run: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/avi/job/avi-regression-latency-650gb-grow-shrink/12/

We see commitlog space used reach very close to the reservation.

**Comment by avikivity** [comment]
Possible cause is that time to seal the memtable is eating away from flush speed.

Possible fix is to increase memtable shares in proportion to distance from commitlog flush point to hard limit point.

Need to check dirty/unspooled memory to understand why they didn't trigger share increase.

**Comment by avikivity** [comment]
Memtable shares do increase:

![image](https://github.com/scylladb/scylladb/assets/1017210/07d50efa-66e1-4a28-988d-ef076cff1280)

**Comment by bhalevy** [comment]
@elcallio please look into this issue

**Comment by mykaul** [comment]
@elcallio - ping

**Comment by elcallio** [comment]
I have literally nothing to add. It is quite possible tablet runs have a different flush behaviour, potentially causing segments to live for to long.

**Comment by bhalevy** [comment]
> I have literally nothing to add. It is quite possible tablet runs have a different flush behaviour, potentially causing segments to live for to long.

Maybe we have more memtables with smaller tablets and that could affect CL segment lifetime.
Should we trigger memtable flush from older segments so we can recycle them, when reaching the soft limit, and do that increasingly aggressively when reaching the hard limit?

**Comment by elcallio** [comment]
We already do initiate flush when reaching about half-way point. From what I can read, the database flush handler (`_tables_metadata.get_table(id).flush(pos)`) should indeed force out all existing memtables when called. Of course, if we already have active flushes going, this will potentially stall while waiting for its turn, so if tables somehow create more in-parallel sstable-to-disk ops, maybe we simply see disk competition...

**Comment by elcallio** [comment]
Would be good to correlate the CL size expansion with sstable writes/memtable flush and general disk throughput...

**Comment by bhalevy** [comment]
@elcallio let's start with reproducing this issue, maybe with artificially high initial tablets count
Cc @regevran

**Comment by bhalevy** [comment]
@avikivity @mykaul should this be a release blocker?

**Comment by kostja** [comment]
I agree with @elcallio this behaviour is most likely by design. The hard limit implementation is not really hard - it simply increase the pressure on memtable flush to keep commit log segments rotating faster. But there is no hard guarantee that the system will not use up all the existing commit log segments - the implementation puts no direct backpressure towards the clients inserting more and more data.

While I agree it woudl be nice to reproduce it, I don't think it's a release blocker.

**Comment by regevran** [comment]
> [@elcallio](https://github.com/elcallio) let's start with reproducing this issue, maybe with artificially high initial tablets count Cc [@regevran](https://github.com/regevran)


The original test is running on a branch that is some 1800 commits behind master.

A [similar test](https://github.com/scylladb/scylla-cluster-tests/blob/d30105c0349e4f32ac23c8dd0aef1ee9c2dbd31d/jenkins-pipelines/performance_staging/scylla-enterprise-perf-regression-latency-650gb-grow-shrink.jenkinsfile#L10) exists but [was removed](https://github.com/scylladb/scylla-cluster-tests/commit/2d0ca5b729f40d571a265fc85b0c395edbae46a3) from the weekly runs on Jul 29. It was [last run](https://jenkins.scylladb.com/view/enterprise/job/scylla-enterprise/job/perf-regression/job/scylla-enterprise-perf-regression-latency-650gb-grow-shrink/) at Jul 28.

There is yet [another test](https://jenkins.scylladb.com/job/scylla-enterprise/job/perf-regression/job/scylla-enterprise-perf-regression-latency-650gb-with-nemesis/24/) which is similar and ran lately. Its results are in [Argus](https://argus.scylladb.com/workspace?state=WyIyYTRkYjlkNS04MGU2LTQzN2UtODg3MS1hNGQ1ZTU0Y2MzNWMiXQ).


@elcallio - please take a look at all of this and see if we can progress.

There is no point in running the original tests, unless we really need them and then we should promote the code first.
(thanks @pehala for supporting).

**Comment by elcallio** [comment]
Looking at the logs for the last run above:

These runs were on i3en.2xlarge instances, right? Those are 64GB machines, which should mean that we have a max total CL size of about the same per instance? Or am I reading the graphs wrong, and the values are per shard? (Should be 8GB in that case). 

In either case, the reserved footprint seems to be ~4GB, with peaks at ~3GB, followed by peaks and lows as flushing happens. We seem to have a higher latency/lower throughput part in first third of the run - maybe a nemesis messing? 

In any case, can someone verify what the actual heap max is here (I can't find it in the logs for the node), because unless it is explicitly set to ~4GB it looks to me like we are in fact peaking the CL size at below 50% of max (?), which is about where we start doing forward flush requests.

**Comment by regevran** [comment]
@avikivity - I would like top close this issue, as I see no point progressing it in its original setup.
If we want to re-test the issue, please suggest a test that can show thew problem.

**Comment by avikivity** [comment]
The test quoted is a standard test. If you want to close the issue, look at a recent run and see if the problem still occurs or is no longer present.

**Comment by regevran** [comment]
This is the [correct test](https://jenkins.scylladb.com/job/scylla-enterprise/job/perf-regression/job/scylla-enterprise-perf-regression-latency-650gb-elasticity/).
Argus [vnodes](https://argus.scylladb.com/tests/scylla-cluster-tests/e4e64580-1d7c-4bcb-a96f-e2ecfc06594a)
Argus [tablets](https://argus.scylladb.com/tests/scylla-cluster-tests/60c3813e-d281-4c33-bad3-d2ff46d9e033)
